Notify other threads before running callbacks

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31713 Notify other threads before running callbacks**

- In case the callbacks are heavy/slow, the other threads should be able to start work on the value of the future after the current thread moves the value and unlock the mutex.
- `completed()` is not inlined. Avoid function call overhead.

Differential Revision: [D5624371](https://our.internmc.facebook.com/intern/diff/D5624371/)

xush6528 committed:
Notify other threads before running callbacks

In case the callbacks are heavy/slow, the other threads should be able to start work on the value of the future after the current thread moves the value and unlock the mutex.

Differential Revision: [D5624371](https://our.internmc.facebook.com/intern/diff/D5624371/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 845598d1:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=845598d17d473c0d75485780fa78bf08ec48db54).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4087895) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78780289) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78780289))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/verbatim-sources/docker_jobs.yml 
Auto-merging .circleci/verbatim-sources/docker_jobs.yml 
CONFLICT (add/add): Merge conflict in .circleci/scripts/python_doc_push_script.sh 
Auto-merging .circleci/scripts/python_doc_push_script.sh 
CONFLICT (add/add): Merge conflict in .circleci/ecr_gc_docker/gc.py 
Auto-merging .circleci/ecr_gc_docker/gc.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_conda.sh 
Auto-merging .circleci/docker/common/install_conda.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "Notify other threads before running callbacks"


- In case the callbacks are heavy/slow, the other threads should be able to start work on the value of the future after the current thread moves the value and unlock the mutex.
- `completed()` is not inlined. Avoid function call overhead.

Differential Revision: [D5624371](https://our.internmc.facebook.com/intern/diff/D5624371/)

[ghstack-poisoned]
